File: src/problem017/solution017.py
names = {
    1: 'one', 2: 'two', 3: 'three', 4: 'four', 5: 'five', 6: 'six', 7: 'seven',
    8: 'eight', 9: 'nine', 10: 'ten', 11: 'eleven', 12: 'twelve',
    13: 'thirteen', 14: 'fourteen', 15: 'fifteen', 16: 'sixteen',
    17: 'seventeen', 18: 'eighteen', 19: 'nineteen', 20: 'twenty',
    30: 'thirty', 40: 'forty', 50: 'fifty', 60: 'sixty', 70: 'seventy',
    80: 'eighty', 90: 'ninety', 100: 'hundred', 1000: 'thousand'
}


def to_english(num):
    words = []
    if num >= 1000:
        words.append('%s %s' % (to_english(num / 1000), names[1000]))
        num %= 1000
    if num >= 100:
        words.append('%s %s' % (to_english(num / 100), names[100]))
        num %= 100
    if num and words:
        words.append('and')
    if num >= 20:
        words.append('%s-' % names[(num / 10) * 10])
        num %= 10
    if num:
        words.append(names[num])
    return ' '.join(words).replace('- ', '-').strip('-')


def main():
    english_nums = [to_english(n) for n in xrange(1, 1000 + 1)]
    print(sum([len([c for c in eng if c.isalpha()]) for eng in english_nums]))


if __name__ == '__main__':
    main()

File: src/problem017/solution017.hs
import Data.Char (isAlpha)

main = print . length . filter isAlpha .
    foldl (\x y -> x ++ "\n" ++ y) "\n" . map toEnglish $ [1..1000]

toEnglish n
    | n >= 1000 = names (n `div` 1000) ++ " " ++ names 1000 ++ " " ++
        toEnglish (n `mod` 1000)
    | n >= 100 && (n `mod` 100 > 0) = names (n `div` 100) ++ " " ++
        names 100 ++ " and " ++ toEnglish (n `mod` 100)
    | n >= 100 = names (n `div` 100) ++ " " ++ names 100 ++ " " ++
        toEnglish (n `mod` 100)
    | n >= 20 && n `mod` 10 > 0 = names ((n `div` 10) * 10) ++ "-" ++
        toEnglish (n `mod` 10)
    | n >= 20 = names ((n `div` 10) * 10) ++ " " ++
        toEnglish (n `mod` 10)
    | n >= 1 = names n
    | n == 0 = ""

names n
    | n == 1 = "one" | n == 2 = "two" | n == 3 = "three" | n == 4 = "four"
    | n == 5 = "five" | n == 6 = "six" | n == 7 = "seven" | n == 8 = "eight"
    | n == 9 = "nine" | n == 10 = "ten" | n == 11 = "eleven"
    | n == 12 = "twelve" | n == 13 = "thirteen" | n == 14 = "fourteen"
    | n == 15 = "fifteen" | n == 16 = "sixteen" | n == 17 = "seventeen"
    | n == 18 = "eighteen" | n == 19 = "nineteen" | n == 20 = "twenty"
    | n == 30 = "thirty" | n == 40 = "forty" | n == 50 = "fifty"
    | n == 60 = "sixty" | n == 70 = "seventy" | n == 80 = "eighty"
    | n == 90 = "ninety" | n == 100 = "hundred" | n == 1000 = "thousand"
